x-image-privacy/word-cloud
# Wordcloud working

This PR proposes changes that result in a "working" wordcloud.

Main changes:
- use `reduce` instead of `map` as it makes more sens
- use the `text-anchor` property set to `middle` to consider the anchor point in the lower center of the word.
- compute only once the width of each element, store it in a `Rectangle` type, that can be manipulated intil the best position is found
- after all words have been placed, fuse the rectangle with the word properties
- add a coef prop to each word that scales the font size (the higher the coef the bigger the word
- method to get and compute element size was changed a bit, but can be improved or simplified
- container has a known size so we do not need to get it from the DOM -> use constants

I experimented with different settings:
- each move is of 3 pixels along the hypotenus
- the move vector is computed as the sum of the difference between centers (no forces)
- the loop stops either when there is a fullCollision (no way to move further), or the total displacement is very small (less than 2px) or after more than 600 iterations
- I tweaked the bounding box to reduce the height of all the elements to pack them more tightly on the y axis (-5) and added a bit of space to let them breath on the x-axis (+10)
- I changed the css so that the text is white when the background is black and inversly depending on the `prefers dark-scheme` property (already included in the vite template, just needed to add a rule for svg text)
- I changed the method to place the word at a random position. I choose a value beween 0° and 360°. So all the words are placed on an imaginary circle that is half the biggest dimension of the container.

Needed:
- tests for all functions
- improve reloading and positioning of words
- compute if word should be placed else where to homogenize the word-cloud
- try to improve parameters (step size, max number of iterations, ...)
- try it on larger data (more words)

<img width="533" alt="Capture d’écran 2023-03-10 à 23 36 46" src="https://user-images.githubusercontent.com/39373170/224441199-fbe5e235-a26e-4c95-8eba-5cd611b8a6ba.png">
